Added zsh and powershell cobra completion targets

Lockszmith opened the pull request:
Per Issue #187, I've copied the code from https://github.com/spf13/cobra/blob/master/shell_completions.md adjusting it to the i18n template that was already used.
I added cobra output only for `powershell` and `zsh`.
It seems that `fish` was introduced after the [fork](/spf13/cobra/pull/983) so I dropped that code.

I only tested the PowerShell version.

didrocks commented on cmd/zsysd/completion.go:
```diff
@@ -11,18 +11,52 @@ import (
-			genBashCompletion(rootCmd, os.Stdout)
+			switch args[0] {
+			case "bash":
+				cmd.Root().GenBashCompletion(os.Stdout)
```
Please use genBashCompletion(rootCmd, os.Stdout) rather, as this is our vendorized version with our fixes for hidden commands I talked about earlier.

didrocks commented on cmd/zsysd/completion.go:
```diff
@@ -11,18 +11,52 @@ import (
 `), prog, prog),
+		DisableFlagsInUseLine: true,
+		ValidArgs:             []string{"bash", "zsh", "fish", "powershell"},
+		Args:                  cobra.ExactValidArgs(1),
```
What do you think about the following:
- Have 0 or 1 arg
- If 0 arg -> detect current shell and execute the expected completion
- If 1 arg: ensuring (as with corba.ExactValidArgs that you have one of the whitelisted list)
- Based on the Shell detected, (I think you can do that with a PreRun), modify completionCmd.Long message to include one of the 3 snippets (bash, zsh or powershell). If not found, prints all 3.

As a sidenote, you added "fish" which is not handled.

didrocks commented on cmd/zsysd/completion.go:
```diff
@@ -11,18 +11,52 @@ import (
 	prog := rootCmd.Name()
 	var completionCmd = &cobra.Command{
-		Use:   "completion",
+		Use:   "completion [bash|zsh|powershell]",
```
Here, you agree (see below comment) that the syntax should have an optional argument :)

didrocks commented on cmd/zsysd/completion.go:
```diff
@@ -11,18 +11,52 @@ import (
 	var completionCmd = &cobra.Command{
-		Use:   "completion",
+		Use:   "completion [bash|zsh|powershell]",
 		Short: i18n.G("Generates bash completion scripts"),
```
Needs, to change, it’s not only bash now! If we do autodetection (see below), we can write: "Generates completion scripts for your current shell"

didrocks commented on cmd/zsysd/completion.go:
```diff
@@ -11,18 +11,52 @@ import (
+Bash:
 
-To configure your bash shell to load completions for each session add to your ~/.bashrc or ~/.profile:
+  $ source <(yourprogram completion bash)
```
please replace yourprogram here (and below) with %s for prog

didrocks reviewed (changes requested):
Good work! Thanks for tackling it!

There are some suggestions how to enhance user experience. I’m happy to give you some help if needed.
In the end, it will need packaging integration (to ship them as part of our deb package). Again, I can do it.

I know this can  be challenging for you to test and ensuring that there is no regression due to build depending on openzfs2.0 (even our CI doesn’t have it as main is for hirsute). I suggest that you revert locally to use go-libzfs binding which is compatible with openzfs1.0:

Basically, rebase on `origin/main` and revert [this commit](https://github.com/ubuntu/zsys/commit/b8ba98dc18f00fb69599f46e5d7ccdf209f27775). That should do it :)

Thanks again, and do not hesite if you need anything :)

Lockszmith commented:
Note to self, don't commit a PR in 2 am after a long day :)

I'll get to fix all of those.

didrocks commented:
:) FYI, the CI fixes are in https://github.com/ubuntu/zsys/pull/191. I’m waiting for jibel to have a quick review, but you can rebase on it. (+ do the revert locally I showed you to get it running on older releases than hirsute)

Lockszmith commented:
I think I did the rebase correctly, please let me know if I messed up anything

Lockszmith commented:
> Basically, rebase on `origin/main` and revert [this commit](https://github.com/ubuntu/zsys/commit/b8ba98dc18f00fb69599f46e5d7ccdf209f27775). That should do it :)

I think I did the rebase correctly, please let me know if I messed up anything
I guess you meant `origin/master`, what I did was:
> Define upstream and rebase master on the b8ba98d commit
```
git remote add upstream https://github.com/ubuntu/zsys.git
git fetch upstream
git checkout master # make sure I'm on the master branch
git rebase upstream/master # This isn't really necessary, but wanted to see that it works.
git revert b8ba98d # The commit that matters
```
> rebase completion on rebased master
```
git checkout completion
git rebase master
```
> Commit changes
```
# Make changes to the code per your suggestions.
git commit -a
git push --force # this pushes to origin (my clone) with all the changes 
```

didrocks commented on cmd/zsysd/completion.go:
```diff
@@ -11,18 +12,58 @@ import (
-		Long: fmt.Sprintf(i18n.G(`To load completion run
+		Use:   "completion [bash|zsh|powershell]",
+		Short: i18n.G("Generates completion scripts"),
+		Long: strings.ReplaceAll(i18n.G(`To load completions:
```
Please use fmt.Sprintf rather than ReplaceAll here, this is more familiar to translators and less up for breakage.

Lockszmith replied:
I used `strings.ReplaceAll` because of the numerous locations the program name appears in the string, instead of manually counting the number of %s instances and providing duplicate parameters at the end.
If you feel that's the wrong approach, I'll switch back to `fmt.Sprintf`

didrocks replied:
Yeah. I got why you were doing this and indeed, named replacement (like in python), would be good. But it’s still better I think to use the standard `fmt.Sprintf` to keep %s for the translators.
OR, IIRC, this is the only string replaced, correct? So you can `strings.ReplaceAll` all `%s` with `prog`? That way, we keep %s for translators without worrying about the underlined mechanic?

Lockszmith replied:
OK, I'll just change the #prog# to %s, np.

didrocks commented on cmd/zsysd/completion.go:
```diff
@@ -11,18 +12,58 @@ import (
+  PS> #prog# completion powershell > #prog#.ps1
+  # and source this file from your PowerShell profile.
+`), "#prog#", prog),
+		DisableFlagsInUseLine: true,
```
We still allow other flags like -vv, no?

Lockszmith replied:
I will remove that.

didrocks commented on cmd/zsysd/completion.go:
```diff
@@ -11,18 +12,58 @@ import (
+  # and source this file from your PowerShell profile.
+`), "#prog#", prog),
+		DisableFlagsInUseLine: true,
+		ValidArgs:             []string{"bash", "zsh", "powershell"},
```
Does this still have an effect with cobra.MaximumNArgs(1)? (any validation done by it)

Lockszmith replied:
Well, it does not deal with the validation, but it does provide completion.
Now, that might not be important (as this is run when completion doesn't work) - so I can remove it.
Validation is handled in the code by the `default` clause of the `switch`.

Let me know what would be preferred here.

didrocks replied:
If this provides completion, this is a good reason enough to keep it :)

Lockszmith replied:
on a related topic, any idea how (from the Run code) I can display the '--help' usage output, I've looked at the cobra docs but can't seem to find the solution.

Lockszmith replied:
I want the usage output to show when the detected shell is blank or unknown

didrocks replied:
Yes! You can get the command from the closure, and call `.Usage()` on it (see https://pkg.go.dev/github.com/spf13/cobra#Command.Usage). I think you can just do it quite elegantly in your switch in the default() case.
Then, after handling the error, just call `os.Exit(1)`.

Is there anything else blocking or did you misread you?

didrocks commented on cmd/zsysd/completion.go:
```diff
@@ -11,18 +12,58 @@ import (
+`), "#prog#", prog),
+		DisableFlagsInUseLine: true,
+		ValidArgs:             []string{"bash", "zsh", "powershell"},
+		Args:                  cobra.MaximumNArgs(1), //cobra.ExactValidArgs(1),
```
you left a comment here :p

didrocks commented on cmd/zsysd/completion.go:
```diff
@@ -11,18 +12,58 @@ import (
+		Args:                  cobra.MaximumNArgs(1), //cobra.ExactValidArgs(1),
 		Run: func(cmd *cobra.Command, args []string) {
-			genBashCompletion(rootCmd, os.Stdout)
+			var arg = "bash"
```
Sorry, I probably didn’t explain it too well. The idea was rather if no arg -> detect current shell (and don’t default, like, for instance, to bash).
I think you can rely on something like TERM env variable for it.

That way, you just type "systemctl completion" and it gives the correct default the user :)
The help should be updated as well to state the default behavior (probably at the top)

Lockszmith replied:
you are right, I didn't get to implement the detection, I think the $SHELL env variable is the way to go here.

didrocks reviewed (changes requested):
We are almost there! I think you didn’t get my last comment on rebasing on the unmerged PR. This PR is now merged so you can (again) rebasing on master now.

I seems you pushed commit b8ba98d, which was only for you to build the project locally and ensuring things are working, but not pushing it! (master is based on openzfs 2.0 and I bet you are using v1 as not running in hirsute?)

See my various comments in the review. Happy to help if you need anything :)

Lockszmith commented:
I think I addressed all issues. 🤞 
Thanks for your patience on this - seems like a trivial task for you, but a learning curve for me.
And thanks for allowing to participate.

didrocks commented on cmd/zsysd/completion.go:
```diff
@@ -11,18 +12,66 @@ import (
 		Run: func(cmd *cobra.Command, args []string) {
-			genBashCompletion(rootCmd, os.Stdout)
+			tmpShell := os.Getenv("SHELL")
+			shell := "unknown"
```
nitpick: I would just here declare and simplify the whole stanza here:
```
var shell string
if len(args) > 0 && args[0] != "" {
	shell = args[0]
} 
shell = filepath.Base(shell)

switch shell…
```

Then, the error message will only print "Shell preset unknown: <value of shell>", I would add something like: "Please specify between bash, zsh, powershell".

Lockszmith replied:
Cleaned up, as for the error message, it is followed by the usage output.
Do you still think this should be changed?

didrocks commented on cmd/zsysd/completion.go:
```diff
@@ -11,18 +12,66 @@ import (
+				cmd.Root().GenPowerShellCompletion(os.Stdout)
+			default:
+				cmd.SilenceUsage = false
+				cmd.PrintErrf("Shell preset unkown: %-36s\n", shell)
```
Out of curiousity, why %-36s?

Lockszmith replied:
Just some leftovers from another string test - I'll remove that

didrocks commented on go.mod:
```diff
@@ -26,4 +26,4 @@ require (
 )
 
-replace github.com/bicomsystems/go-libzfs => github.com/ubuntu/go-libzfs v0.2.2-0.20210126140656-217c247dccce
+replace github.com/bicomsystems/go-libzfs => github.com/ubuntu/go-libzfs v0.2.2-0.20200410091938-105a6f043895
```
This shouldn’t be in the diff, no? (that was for local testing < hirsute)

Lockszmith replied:
probably, didn't notice it - will revert.

didrocks commented on cmd/zsysd/completion.go:
```diff
@@ -11,18 +12,66 @@ import (
 
-. <(%s completion)
-`), prog, prog),
+  # To load completions for each session, execute once:
```
I think I mentined it previously. bu I don’t remember, I think the top should mention that we try to automate the shell detection and that the argument of the shell variable shouldn’t be needed.

Lockszmith replied:
I'll add some additional description

didrocks reviewed (commented):
Some nitpicks + revert of some changes, but overall this looks good :)
No worry for the patience, thanks to *you* for working on this :) In addition, if this can get you some Go skills, bonus point :)

You are correct to mark conversation as resolved if you feel those are fixed, there are just some small changes needed. Hoping this is clear enough :)

Lockszmith commented:
Commit is in.

didrocks commented:
Excellent, that looks good to me! I will probably rephrase one or two strings in the near future, but the whole logic is there and code looks good :)

Thanks again for tackling this. I hope you had fun with it! And do not hesitate to continue your good work on other features (I heared the embedeed doc stuff? :p)
